File: python/asyncpg/src/alternatives/no_connection_pool/example_with_no_connection_pool.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

import asyncio
import os

import aurora_dsql_asyncpg as dsql


async def create_connection(cluster_user, cluster_endpoint):
    ssl_cert_path = "./root.pem"
    if not os.path.isfile(ssl_cert_path):
        raise FileNotFoundError(f"SSL certificate file not found: {ssl_cert_path}")

    conn_params = {
        "user": cluster_user,
        "host": cluster_endpoint,
        "ssl": "verify-full",
        "sslrootcert": ssl_cert_path,
    }

    # Make a connection to the cluster
    conn = await dsql.connect(**conn_params)

    if cluster_user == "admin":
        schema = "public"
    else:
        schema = "myschema"

    try:
        await conn.execute(f"SET search_path = {schema};")
    except Exception as e:
        await conn.close()
        raise e

    return conn


async def exercise_connection(conn):
    await conn.execute(
        """
        CREATE TABLE IF NOT EXISTS owner(
            id uuid NOT NULL DEFAULT gen_random_uuid(),
            name varchar(30) NOT NULL,
            city varchar(80) NOT NULL,
            telephone varchar(20) DEFAULT NULL,
            PRIMARY KEY (id))
            """
    )

    # Clean up any data from previous runs to avoid assertion failures
    await conn.execute("DELETE FROM owner WHERE name = $1", "John Doe")

    # Insert some rows
    await conn.execute(
        "INSERT INTO owner(name, city, telephone) VALUES($1, $2, $3)",
        "John Doe",
        "Anytown",
        "555-555-1999",
    )

    row = await conn.fetchrow("SELECT * FROM owner WHERE name=$1", "John Doe")

    # Verify the result we got is what we inserted before
    assert row[0] is not None
    assert row[1] == "John Doe"
    assert row[2] == "Anytown"
    assert row[3] == "555-555-1999"

    # Clean up the table after the example. If we run the example again
    # we do not have to worry about data inserted by previous runs
    await conn.execute("DELETE FROM owner WHERE name = $1", "John Doe")


async def main():
    conn = None
    try:
        cluster_user = os.environ.get("CLUSTER_USER", None)
        assert cluster_user is not None, "CLUSTER_USER environment variable is not set"

        cluster_endpoint = os.environ.get("CLUSTER_ENDPOINT", None)
        assert cluster_endpoint is not None, "CLUSTER_ENDPOINT environment variable is not set"

        conn = await create_connection(cluster_user, cluster_endpoint)
        await exercise_connection(conn)
    finally:
        if conn is not None:
            await conn.close()

    print("Connection exercised successfully")


if __name__ == "__main__":
    asyncio.run(main())

File: javascript/postgres-js/src/alternatives/no_connection_pool/example_with_no_connection_pool.js
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import assert from "node:assert";
import { auroraDSQLPostgres } from "@aws/aurora-dsql-postgresjs-connector";

const ADMIN = "admin";
const PUBLIC = "public";
const NON_ADMIN_SCHEMA = "myschema";

function getConnection(clusterEndpoint, user) {
  return auroraDSQLPostgres({
    host: clusterEndpoint,
    user: user,
    retry: true,
    // Other DSQL options:
    // region: 'us-east-1',
    // profile: awsProfile,
    // tokenDurationSecs: 30,
    // customCredentialsProvider: credentialsProvider,
    //
    // Other Postgres.js settings are also valid here, see Postgres.js documentation for more information
    // https://github.com/porsager/postgres#all-postgres-options
  });
}

async function example() {
  const clusterEndpoint = process.env.CLUSTER_ENDPOINT;
  assert(clusterEndpoint, "CLUSTER_ENDPOINT environment variable is not set");
  const user = process.env.CLUSTER_USER;
  assert(user, "CLUSTER_USER environment variable is not set");

  let client;
  try {
    client = getConnection(clusterEndpoint, user);
    const schema = user === ADMIN ? PUBLIC : NON_ADMIN_SCHEMA;

    // Note that due to connection pooling, we cannot execute 'set search_path=myschema'
    // because we cannot assume the same connection will be used.
    await client`CREATE TABLE IF NOT EXISTS ${client(schema)}.owner
                 (
                     id        UUID PRIMARY KEY DEFAULT gen_random_uuid(),
                     name      VARCHAR(30) NOT NULL,
                     city      VARCHAR(80) NOT NULL,
                     telephone VARCHAR(20)
                 )`;

    // Transactional write with OCC retry
    await client.begin(async (tx) => {
      await tx`INSERT INTO ${tx(schema)}.owner(name, city, telephone)
               VALUES ('John Doe', 'Anytown', '555-555-0150')`;
    });

    // Check that data is inserted by reading it back
    const result = await client`SELECT id, city
                                FROM ${client(schema)}.owner
                                where name = 'John Doe'`;
    assert.deepEqual(result[0].city, "Anytown");
    assert.notEqual(result[0].id, null);

    // Delete data we just inserted
    await client`DELETE
                 FROM ${client(schema)}.owner
                 where name = 'John Doe'`;

    console.log("Completed successfully");
  } catch (error) {
    console.error(error);
    throw error;
  } finally {
    await client?.end();
  }
}

export { example };
